### PR TITLE
use staging location in sync-granule

### PR DESCRIPTION
**Pull request by sethvincent**
**Summary:** Updates sync-granule task to use staging location

Addresses [CUMULUS-456](https://bugs.earthdata.nasa.gov/browse/CUMULUS-456)

## Changes

- Updates sync-granule task to use staging location
  - uses staging location
  - updates schemas
  - adds and updates tests

## Test Plan
Things that should succeed before merging.

- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Update CHANGELOG
- [ ] Run `./bin/eslint-ratchet` and verify that eslint errors have not increased
  - [ ] Commit `.eslint-ratchet-high-water-mark` if the score has improved

### CODE REVIEW

File: packages/ingest/granule.js [hunk 1]
```diff
@@ -50,45 +58,50 @@ class Discover {
...
+  * @returns {Array} an array of granule objects
+  **/
   async discover() {
     // get list of files that matches a given path
     const updatedFiles = (await this.list())
       .map((file) => this.setGranuleInfo(file))
       .filter((file) => file);
 
-    return await this.findNewGranules(updatedFiles);
+    return this.findNewGranules(updatedFiles);
```
Comment (laurenfrederick) [inline packages/ingest/granule.js]: Why did you remove the `await` here?

File: packages/ingest/granule.js [hunk 2]
```diff
@@ -235,37 +290,64 @@ class Granule {
...
-  async _hash(type, tempFile) {
-    const options = { algorithm: type };
+  /**
+  * Get hash of file
+  *
+  * @param {string} algorithm - algorithm to use for hash,
+  * any algorithm accepted by node's `crypto.createHash`
+  * https://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm_options
+  * @param {string} filepath - filepath of file to checksum
+  * @returns {Promise} checksum value calculated from file
```
Comment (laurenfrederick) [inline packages/ingest/granule.js]: This should be `{Promise<type>}` with the type that the promise resolves to

File: packages/ingest/granule.js [hunk 3]
```diff
@@ -235,37 +290,64 @@ class Granule {
...
-    );
+      }));
   }
 
-  async enableDuplicateHandling(bucket) {
+  /**
+   * Enable versioning on an s3 bucket
+   *
+   * @param {string} bucket - s3 bucket name
+   * @returns {Promise} promise that resolves when bucket versioning is enabled
```
Comment (laurenfrederick) [inline packages/ingest/granule.js]: You removed the return so it's not return anything.

File: packages/ingest/url-path-template.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+* define the path of a file based on the metadata of a granule
+*
+* @param {string} pathTemplate - the template that defines the path,
+* using `{}` for string interpolation
+* @param {Object} context - the metadata used in the template
+* @returns {string} - the url path for the file
+**/
+module.exports = function urlPathTemplate(pathTemplate, context) {
```
Comment (laurenfrederick) [inline packages/ingest/url-path-template.js]: Write unit tests for this function

File: tasks/sync-granule/index.js [hunk 5]
```diff
@@ -69,6 +69,7 @@ exports.syncGranule = function syncGranule(event) {
     buckets,
     collection,
     provider,
+    fileStagingDir,
```
Comment (laurenfrederick) [inline tasks/sync-granule/index.js]: There is a default here, right? It might help to put a comment in the config.json what the default is and maybe in here because upon first glance it looks like it should be required.

File: .eslintrc.json [hunk 6]
```diff
@@ -51,6 +51,7 @@
     "generator-star-spacing": "off",
     "import/no-extraneous-dependencies": "off",
     "import/newline-after-import": "off",
+    "class-methods-use-this": "off",
```
Comment (yjpa7145) [inline .eslintrc.json]: Where did you run into problems with this rule?

File: packages/ingest/granule.js [hunk 7]
```diff
@@ -50,45 +58,50 @@ class Discover {
...
-   * @returns {object} Updated file with granuleId, bucket and path information
+   *
+   * @param {Object} file - the file object
+   * @returns {Object} - Updated file with granuleId, bucket and path information
    */
-  setGranuleInfo(_file) {
-    let granuleId;
-    const file = _file;
+  setGranuleInfo(file) {
+    const _file = file;
```
Comment (yjpa7145) [inline packages/ingest/granule.js]: What is this line doing?

Comment (scisco) [reply packages/ingest/granule.js]: I think this line gets around our very irritating eslint rule: `no-param-reassign`

File: packages/ingest/granule.js (hunk 1 above)
Comment (yjpa7145) [reply packages/ingest/granule.js]: `this.findNewGranules()` returns a Promise.  There is not need to wait for that promise to resolve before returning from the discover function.  Not having await here is correct.

File: packages/ingest/granule.js (hunk 3 above)
Comment (scisco) [reply packages/ingest/granule.js]: It still returns a promise even there is no `return` because it is an async function

File: packages/ingest/granule.js [hunk 8]
```diff
@@ -108,17 +121,26 @@ class Discover {
...
+
     const t = await Promise.all(checkFiles);
     const newFiles = t.filter((f) => f);
 
     // reorganize by granule
     const granules = {};
     newFiles.forEach((_f) => {
       const f = _f;
-      const granuleId = f.granuleId;
+      const { granuleId } = f;
```
Comment (yjpa7145) [inline packages/ingest/granule.js]: One of our recent PRs disabled the `prefer-destructuring` rule in eslint, which I'm pretty sure is why this was changed.  I'm open to discussion on it, but I think that `const granuleId = f.granuleId` is easier to read than ` const { granuleId } = f`

File: packages/ingest/granule.js (hunk 3 above)
Comment (yjpa7145) [reply packages/ingest/granule.js]: Good catch.

File: packages/ingest/granule.js (hunk 1 above)
Comment (scisco) [reply packages/ingest/granule.js]: yeah, await is redundant when you are just passing a promise from another function

File: packages/ingest/url-path-template.js (hunk 4 above)
Comment (sethvincent) [reply packages/ingest/url-path-template.js]: I'm going to move this function and unit tests to a different pr as it's not actually used in these changes.

File: .eslintrc.json (hunk 6 above)
Comment (sethvincent) [reply .eslintrc.json]: Here's an example: https://github.com/cumulus-nasa/cumulus/blob/master/packages/ingest/granule.js#L290

cmrjs package has a few examples: https://github.com/cumulus-nasa/cumulus/blob/master/packages/cmrjs/index.js

ingest/protocol has examples: https://github.com/cumulus-nasa/cumulus/blob/master/packages/ingest/protocol.js

I started out having eslint ignore specific functions like those ^, but found enough instances of it that I gave up.

I think the assertion that rule is making is that those methods should be functions that are not attached to classes. I agree with that assertion, but this code base has enough class methods that don't use `this` that it doesn't seem like a good use of effort to refactor them all, mostly because the examples I'm finding aren't particularly useful separated from the classes they're attached to.

File: packages/ingest/granule.js (hunk 7 above)
Comment (sethvincent) [reply packages/ingest/granule.js]: Yeah, `no-param-reassign` is a regular irritation. It seems like a good practice to follow in general, but we have a lot of code that's meant to take an object, make changes, and return the changed object, so it gets annoying.

File: packages/ingest/granule.js (hunk 8 above)
Comment (sethvincent) [reply packages/ingest/granule.js]: Totally agree, I'm very happy to not use destructuring in cases like this.

File: tasks/sync-granule/index.js (hunk 5 above)
Comment (sethvincent) [reply tasks/sync-granule/index.js]: :+1:

File: packages/ingest/granule.js (hunk 3 above)
Comment (laurenfrederick) [reply packages/ingest/granule.js]: Ahhh so much to learn!
